compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```
@@ -292,14 +321,124 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
+                // FIXME: clean this up, likely by permitting array patterns when matching on slices
+                ty::Array(elem_ty, _) if elem_ty == tcx.types.u8 => PatKind::Constant { value: cv },
+                // Cannot merge this with the catch all branch below, because the `const_deref`
+                // changes the type from slice to array, and slice patterns behave differently from
+                // array patterns.
+                ty::Slice(..) => {
+                    let old = self.behind_reference.replace(true);
+                    let array = tcx.deref_const(self.param_env.and(cv));
+                    let val = PatKind::Deref {
+                        subpattern: Pat {
+                            kind: Box::new(PatKind::Slice {
+                                prefix: tcx
+                                    .destructure_const(param_env.and(array))
+                                    .fields
+                                    .iter()
+                                    .map(|val| self.recur(val, false))
+                                    .collect(),
+                                slice: None,
+                                suffix: vec![],
+                            }),
+                            span,
+                            ty: pointee_ty,
+                        },
+                    };
+                    self.behind_reference.set(old);
+                    val
+                }
+                // Backwards compatibility hack: support references to non-structural types.
+                // We'll lower
+                // this pattern to a `PartialEq::eq` comparison and `PartialEq::eq` takes a
+                // reference. This makes the rest of the matching logic simpler as it doesn't have
+                // to figure out how to get a reference again.
+                ty::Adt(..) if !self.type_marked_structural(pointee_ty) => {
+                    PatKind::Constant { value: cv }
```

Didn't this used to be the old `ty::Ref(_, adt_ty, _) if adt_ty.is_adt() && !self.type_marked_structural(adt_ty)` case, which had a *hard error*? How is it silently accepted now?

---

Uff, good catch. This is now a hard error again

---

If this is subtle and not covered by tests so far, is it possible to add a test?

---

it was covered by tests :sweat_smile:

---

Oh... I didn't realize this PR had a red checkmark before.^^